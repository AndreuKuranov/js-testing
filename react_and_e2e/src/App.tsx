// import React, { ReactElement, useEffect, useState } from 'react';
// import './App.css';

// const App = (): ReactElement => {
//   const [data, setData] = useState<null | {}>(null);
//   const [toggle, setToggle] = useState<boolean>(false);
//   const [value, setValue] = useState<string>('');

//   const onClick = (): void => setToggle(prev => !prev);

//   useEffect(() => {
//     setTimeout(() => {
//       setData({})
//     }, 100)
//   }, [])

//   return (
//     <div className="App">
//       <h1 data-testid="value-elem">{value}</h1>
//       {toggle === true && <div data-testid="toggle-elem">toggle</div>}
//       {data && <div style={{color: 'red'}}>data</div>}
//       <h1 className="h1">test</h1>
//       <button data-testid="toggle-btn" onClick={onClick}>click</button>
//       <input onChange={e => setValue(e.target.value)} type="text" placeholder='value...'/>
//     </div>
//   );
// }

// export default App;

// -------------------------------------------------------------------------

import React from 'react';
import { Routes, Route, Link } from "react-router-dom";
import MainPage from './pages/MainPage';
import AboutPage from './pages/AboutPage';
import Users from './components/users/Users';
import ErrorPage from './pages/ErrorPage';
import UserDetailsPage from './pages/UserDetailsPage';
import AppRouter from './router/AppRouter';

const App = () => {
  return (
    <div>
      <Link to="/" data-testid="main-link">main</Link>
      <Link to="/about" data-testid="about-link">about</Link>
      <Link to="/users" data-testid="users-link">users</Link>
      <AppRouter />
    </div>
  );
};

export default App;
