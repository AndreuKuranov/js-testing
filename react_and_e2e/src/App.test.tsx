/* eslint-disable testing-library/prefer-presence-queries */
/* eslint-disable testing-library/await-async-query */
/* eslint-disable testing-library/no-debugging-utils */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';

describe('test app', () => {
  it('renders learn react link', () => {
    render(<App />);
    const test = screen.getByText(/test/i);
    const btn = screen.getByRole('button');
    const input = screen.getByPlaceholderText(/value.../i);
    expect(test).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
    // screen.debug();
  });

  it('renders learn react link 2', () => {
    render(<App />);
    const test = screen.queryByText(/test2/i);
    expect(test).toBeNull();
  });

  it('renders learn react link 3', async () => {
    render(<App />);
    // screen.debug();
    const test = await screen.findByText(/data/i);
    expect(test).toBeInTheDocument();
    expect(test).toHaveStyle({color: 'red'});
    // screen.debug();
  });

  it('click event', async () => {
    render(<App />);
    const btn = screen.getByTestId('toggle-btn');
    expect(screen.queryByTestId('toggle-elem')).toBeNull();
    fireEvent.click(btn); // событие клика
    expect(screen.queryByTestId('toggle-elem')).toBeInTheDocument();
    fireEvent.click(btn);
    expect(screen.queryByTestId('toggle-elem')).toBeNull();
  });

  it('input event', async () => {
    render(<App />);
    const input = screen.getByPlaceholderText(/value.../i);
    expect(screen.queryByTestId('value-elem')).toContainHTML(''); // ожидается пустая строка
    // Искуственное событие
    // fireEvent.input(input, {
    //   target: {value: '123123'}
    // }); 
    // Близко к пользователю, обрабатываются события нажатия клавиш и тд
    userEvent.type(input, '123123')
    expect(screen.queryByTestId('value-elem')).toContainHTML('123123');
  });
})


