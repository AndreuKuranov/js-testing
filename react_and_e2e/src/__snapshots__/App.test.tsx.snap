// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test app renders learn react link 1`] = `
<input
  placeholder="value..."
  type="text"
/>
`;
